iroh-util/src/lock.rs

```diff
@@ -1,79 +1,304 @@
-        // of FileGuard releases the underlying lock.
-        if let Ok(file) = File::create(&self.path) {
-            file_guard::try_lock(&file, Lock::Exclusive, 0, 1).is_err()
+        // path exists, examine lock PID
+        let pid = read_lock(&self.path)?;
+        let running =
+            process_is_running(pid).map_err(|e| LockError::Uncategorized { source: e })?;
+        if running {
+            Ok(pid)
         } else {
-            false
+            Err(LockError::ZombieLock(self.path.clone()))
         }
     }
 
     /// Try to acquire a lock for this program.
-    pub fn acquire(&mut self) -> Result<()> {
-        let file = Rc::new(File::create(&self.path)?);
+    pub fn acquire(&mut self) -> Result<(), LockError> {
+        match self.is_locked() {
+            Ok(false) => self
+                .write()
+                .map_err(|e| LockError::Uncategorized { source: anyhow!(e) }),
+            Ok(true) => Err(LockError::Locked(self.path.clone())),
+            Err(e) => match e {
+                LockError::CorruptLock(_) => {
+                    // overwrite corrupt locks
+                    self.write().map_err(|e| LockError::Uncategorized {
+                        source: anyhow!("{}", e),
+                    })
+                }
+                e => Err(LockError::Uncategorized {
+                    source: anyhow!("{}", e),
+                }),
+            },
+        }
+    }
+
+    fn write(&mut self) -> AnyhowResult<()> {
+        // create lock. ensure path to lock exists
+        std::fs::create_dir_all(&crate::iroh_data_root()?)?;
+        let mut file = File::create(&self.path)?;
+        let pid = sysinfo::get_current_pid().unwrap();
+        file.write_all(pid.to_string().as_bytes())?;
+        self.lock = Some(pid);
+        Ok(())
+    }
 
-        file_guard::lock(file, Lock::Exclusive, 0, 1)
-            .map(|lock| self.lock = Some(lock))
-            .map_err(|err| err.into())
+    pub fn destroy_without_checking(&self) -> AnyhowResult<()> {
+        std::fs::remove_file(&self.path).map_err(|e| e.into())
     }
 }
 
+impl Drop for ProgramLock {
+    fn drop(&mut self) {
+        if self.lock.is_some() {
+            if let Err(err) = std::fs::remove_file(&self.path) {
+                warn!("removing lock: {}", err);
+            }
+        }
+    }
+}
+
+fn process_is_running(pid: Pid) -> AnyhowResult<bool> {
+    // existentialism is sometimes counterproductive
+    let this_pid = sysinfo::get_current_pid().unwrap();
+    if pid == this_pid {
+        return Ok(true);
+    }
+
+    // TODO(b5) - docs say we shouldn't be allocating on each call like this:
+    // https://docs.rs/sysinfo/0.26.5/sysinfo/index.html#usage
+    // I'm suspicious this pattern might be alright. Seems the underlying lib
+    // doesn't do much hydrating, but System may be a large memory allocation
+    let mut s = System::new();
```

Comment:
`process_is_running` should just be a method on `ProgrammLock`, which can lazily store a `System`